test/fx/test_dce_pass.py

```diff
@@ -221,3 +222,25 @@ def forward(self, a: torch.Tensor) -> torch.Tensor:
 
         # %add_out node should not be removed because it has side effects.
         self._run_dce_and_test(TestModule(), expect_dce_changes=False, custom=True)
+
+    def test_keep_collectives(self):
+        """
+        Test that DCE doesn't remote collective ops even the results are not used.
+        """
+
+        class TestModule(torch.nn.Module):
+            def __init__(self, *args, **kwargs) -> None:
+                super().__init__(*args, **kwargs)
+                os.environ['MASTER_ADDR'] = '127.0.0.1'
+                os.environ['MASTER_PORT'] = '29500'
+                torch.distributed.init_process_group("gloo", rank=0, world_size=1)
+
+            def forward(self, a: torch.Tensor, b: torch.Tensor, c: torch.Tensor) -> torch.Tensor:
+                d = torch.ops.aten.mul.Tensor(a, b)
+                e = torch.ops.aten.mul.Tensor(a, c)
+                future = torch.ops._c10d_functional.all_reduce.default(e, "sum", "0")
+                synced_e = torch.ops._c10d_functional.wait_tensor.default(future)  # synced_e is not used
+                return d
+
+        # collective nodes should not be removed because they have side effects.
+        self._run_dce_and_test(TestModule(), expect_dce_changes=False, custom=False)
```

Comment:
Instead of calling `init_process_group()` in `TestModule.__init__`, would you like to do something like the following outside of the `TestModule`:
```python
from torch.testing._internal.distributed.fake_pg import FakeStore

...
        torch.distributed.init_process_group(
            backend="fake",
            world_size=2,
            rank=0,
            store=FakeStore(),
        )
        self._run_dce_and_test(...)
        dist.destroy_process_group()
```

Reply:
changed